vod/subtitle/dfxp_format.c

```diff
@@ -481,193 +504,341 @@ dfxp_append_string(u_char* p, u_char* s)
+		if (vod_strcmp(n->name, (u_char *) syle_containers[i]) == 0)
+			return 1;
+
+	return 0;
+}
+
+// dfxp_add_textflags ORs any decorations flags founds in the xmlNode
+// in the flag bits and returns flag:
+static char
+dfxp_add_textflags(xmlNode* n, char flag)
+{ 
+	for (int i = 0; decorationtab[i].name != NULL; i++)
+		flag |= dfxp_has_attr_value(n, decorationtab[i].attr, decorationtab[i].name) << i;
+
+	return flag;
+}
+
+// dfxp_parse_style extracts alignment, display, and decoration
+// attributes and applies it to the style. Its search method is
+// lazy per category. So the first display alignment, text alignment
+// and decoration will be the one used.
+//
+// it searches the pre-declared static style tables
+static style*
+dfxp_parse_style(xmlNode* n, style *s)
+{ 
+	for (int i = 0; regiontab[i].id != NULL; i++) {
+		if (dfxp_has_attr_value(n, "region", regiontab[i].id)){
+			// TODO(as): clearly, we can check if it has a region attr at all
+			// so we dont have to run this loop over and over again if that
+			// tag doesn't exist
+			
+			// TODO(as) should the parent merge with the region? or just
+			// the level nodes and children?
+			*s = regiontab[i].style;
+			break;
 		}
+	}
 
-		switch (cur_node->type)
-		{
-		case XML_TEXT_NODE:
-		case XML_CDATA_SECTION_NODE:
-			result += vod_strlen(cur_node->content);
+	for (int i = 0; textaligntab[i].name != NULL; i++){
+		if (dfxp_has_attr_value(n, textaligntab[i].attr, textaligntab[i].name)){
+			s->flag.text = i;
 			break;
+		}
+	}
 
-		case XML_ELEMENT_NODE:
-			if (vod_strcmp(cur_node->name, DFXP_ELEMENT_BR) == 0)
-			{
-				result++;		// \n
-				break;
-			}
+	for (int i = 0; displayaligntab[i].name != NULL; i++) {
+		if (dfxp_has_attr_value(n, displayaligntab[i].attr, displayaligntab[i].name)){
+			s->flag.display = i;
+			break;
+		}
+	}
 
-			if (vod_strcmp(cur_node->name, DFXP_ELEMENT_SPAN) != 0 ||
-				cur_node->children == NULL ||
-				node_stack_pos >= vod_array_entries(node_stack))
-			{
-				break;
-			}
+	s->flag.decoration |= dfxp_add_textflags(n, s->flag.decoration);
 
-			node_stack[node_stack_pos++] = cur_node->next;
-			cur_node = cur_node->children;
-			continue;
+	return s;
+}
 
-		default:
-			break;
+// dfxp_append_tag applies the HTML-like text decoration
+// tag to p, according to the difference between the flag bits
+// and the parent flag bits, and returns p. 
+//
+// If close is non-zero, close tags (i.e., </b>) are applied instead
+// of open tags.
+//
+static u_char* 
+dfxp_append_tag(u_char* p, char flag, char parentflag, int close, u_char* appendfunc (u_char*, u_char*))
+{
+	// NOTE(as): we only want to append an open or close tag here
+	// if the child has something the parent doesn't. This ensures
+	// we don't have redundant tags across nodes and their ancestors.
+	flag &= (flag ^ parentflag);
+
+	if (flag == 0)
+	{
+		return p;
+	}
+
+	if (close == 0)
+	{
+		for (int i = 0; decorationtab[i].name != NULL; i++)
```

Comment:
quick q. I took a quick pass at this and nothing jumps out at me. One thing I was curious about was I see a few vars like `decortiontab` declared completely lowercase instead of a camel case or `_` and was wondering if it was a c styling convention of some kind. It's been a while so genuinely unsure. 

I found this doc 

https://www.doc.ic.ac.uk/lab/cplus/cstyle.html#:~:text=poorly%2Dorganized%20code.-,File%20Naming%20Conventions,lower%2Dcase%20letters%20and%20numbers.

how do you keep up to date with C? Its been a while for me.

Reply:
Unfortunately, modern C code doesn't take a strong position on styling, most likely due to a lack of tooling that enforces a particular style. My strategy is to write it the way I'm used to (Plan 9 C dialect) and then convert it into the style used by the repository before merging it.

The vod module and the parent nginx code actually differ in coding style as well. The K&R C book serves as a good reference for what a sane styling can look like in C sources. Linux sources often prefer underscores. Windows camel case. Plan 9 just lower-cases the names in most cases, but the names there are usually short enough to not be confusing to parse. 

Given that the vod module uses underscores everywhere, it probably makes the most sense to use them to separate words in globals.

Reply:
> how do you keep up to date with C? Its been a while for me.

Luckily, C doesn't evolve much at the core like the more-complex C++. I highly recommend K&R as a reference, along with "The Practice of Programming" (Kernighan and Pike) for C data structures and algorithms. The latest C standards aren't too relevant to modern day-by-day work, so the old literature has stood the test of time very well.

Some well organized and well written C for the Plan 9 operating system:
https://github.com/0intro/plan9/tree/master/sys/src